research/conditional/utils/layer_manager.py

```diff
@@ -129,20 +129,28 @@ def log_heavy(self):
 
 
 @contextmanager
-def measure_time(obj: LoggingLayer, instruction_name: str):
+def measure_time(layer: LoggingLayer, instruction_name: str):
     """
     This simple context manager is used to measure the time of a block of code.
     Args:
-        obj: The LoggingLayer object that will be used to cache the time.
+        layer: The LoggingLayer object that will be used to cache the time.
         instruction_name: The name of the instruction that is being measured.
     """
-    if obj.logging_switch and torch.cuda.is_available():
-        torch.cuda.synchronize()
-    start_time = time.time()
-
+    if layer.logging_switch:
+        if torch.cuda.is_available():
+            start = torch.cuda.Event(enable_timing=True)
+            end = torch.cuda.Event(enable_timing=True)
+            start.record()
```

Comment:
don't we need cuda.synchronize before recording? or maybe we don't need it after end.record? I'm not sure how does it work exactly.

Reply:
CUDA and CPU are async, and all we need to do is send time markers to CUDA before and aftre the operation we want to measure. Afterwards, we force the CPU to wait until the CUDA operations are finished and synced to the CPU, so we get actual values of the time markers.

Read more here: https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964/11?u=szymon_antoniak

Reply:
ok, nice